streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/task/CheckpointProcessor.java

```diff
@@ -52,30 +62,31 @@ public class CheckpointProcessor {
 
   @Autowired private SavePointService savePointService;
 
-  public void process(Long appId, CheckPoints checkPoints) {
-    CheckPoints.Latest latest = checkPoints.getLatest();
-    if (latest == null || latest.getCompleted() == null) {
-      return;
-    }
-    CheckPoints.CheckPoint checkPoint = latest.getCompleted();
+  public void process(Long appId, @Nonnull CheckPoints checkPoints) {
+    checkPoints.getLatestCheckpoint().forEach(checkPoint -> process(appId, checkPoint));
+  }
+
+  private void process(Long appId, @Nonnull CheckPoints.CheckPoint checkPoint) {
     Application application = applicationService.getById(appId);
     CheckPointStatus status = checkPoint.getCheckPointStatus();
 
     if (CheckPointStatus.COMPLETED.equals(status)) {
-      String cacheId = appId + "_" + application.getJobId();
-      Long latestId =
-          checkPointCache.get(
-              cacheId,
-              key -> {
-                SavePoint savePoint = savePointService.getLatest(appId);
-                return Optional.ofNullable(savePoint).map(SavePoint::getChkId).orElse(null);
-              });
-
-      if (latestId == null || latestId < checkPoint.getId()) {
-        saveSavepoint(checkPoint, application);
+      String cacheId = getCacheIdForCheckpoint(appId, application.getJobId());
+      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
+      if (shouldStoreAsSavepoint(checkPoint, latestChkId)) {
+        savepointedCache.put(checkPoint.getExternalPath(), DEFAULT_FLAG_BYTE);
+        if ((latestChkId == null || checkPoint.getId() > latestChkId)) {
+          checkPointCache.put(cacheId, checkPoint.getId());
```

Comment:
If `savepointedCache` is necessary, why do  you store the savepoint to checkPointCache? Could it just be stored in `savepointedCache`?

In fact, the logic is a bit more complicated now, it's better to simplify it. It reduces the chance of bugs and is easier for other developers to understand.

Reply:
If chk-id of savepoint is bigger than the latest chk-id in the checkpoint cache, the  chk-id of savepoint is useful for the checkpoints store.

Reply:
The checkpoint works well if we don't store savepoint into the check point cache.

When start job from cp or sp, we can sort them by trigger time to choose the latest one. Is there any bug?